internal/pkg/service/common/servicectx/servicectx.go

```diff
@@ -144,64 +174,66 @@ func NewForTest(t *testing.T, opts ...Option) *Process {
 
 // Shutdown triggers termination of the Process.
 func (v *Process) Shutdown(ctx context.Context, err error) {
+	ctx = context.WithValue(ctx, ctxShutdownReasonKey, err) // see ShutdownReason function
+
 	v.lock.Lock()
-	if v.terminating {
-		v.lock.Unlock()
-		return
-	}
-	v.terminating = true
-	v.lock.Unlock()
-
-	v.logger.Infof("exiting (%v)", err)
-	v.errCh <- err
-	close(v.errCh)
-	v.shutdownCtxCh <- ctx
-	close(v.shutdownCtxCh)
-}
+	defer v.lock.Unlock()
 
-func (v *Process) WaitForShutdown() {
-	// Wait for signal
-	_, ok := <-v.errCh
-	if !ok {
-		// Channel is closed, the process is already terminating, wait for completion
-		v.wg.Wait()
+	select {
+	case <-v.terminating:
 		return
+	default:
+		close(v.terminating)
+		v.logger.Infof("exiting (%v)", err)
+		v.shutdown <- ctx
+		close(v.shutdown)
 	}
-
-	// Send cancellation signal to the goroutines.
-	v.cancel()
-
-	// Wait for all operations
-	v.wg.Wait()
-
-	v.logger.Info("exited")
 }
 
 // UniqueID returns unique process ID, it consists of hostname and PID.
 func (v *Process) UniqueID() string {
 	return v.uniqueID
 }
 
-// Add an operation.
-// The Process is graceful terminated when all operations are completed.
-// The ctx parameter can be used to wait for the service termination.
-// The errCh parameter can be used to stop the service with an error.
-func (v *Process) Add(operation func(context.Context, ShutdownFn)) {
-	v.wg.Add(1)
-	go func() {
-		defer v.wg.Done()
-		operation(v.ctx, v.Shutdown)
-	}()
+// Add a root operation.
+// The operation can terminate the entire process by calling the shutdown callback with an error.
+func (v *Process) Add(operation func(ShutdownFn)) {
+	v.lock.Lock()
+	defer v.lock.Unlock()
+
+	select {
+	case <-v.terminating:
+		v.logger.Errorf(`cannot Add operation: the Process is terminating`)
+	default:
+		v.wg.Add(1)
+		go func() {
+			defer v.wg.Done()
+			operation(v.Shutdown)
+		}()
+	}
 }
 
 // OnShutdown registers a callback that is invoked when the process is terminating.
 // Graceful shutdown waits until the callback has finished.
-// Callbacks are invoked sequentially, in LIFO order, see the New function.
+// Callbacks are invoked sequentially, in LIFO order.
 func (v *Process) OnShutdown(fn OnShutdownFn) {
 	v.lock.Lock()
-	if v.terminating {
-		v.logger.Errorf(`cannot register OnShutdown callback: the process is terminating`)
+	defer v.lock.Unlock()
+
+	select {
+	case <-v.terminating:
+		v.logger.Errorf(`cannot register OnShutdown callback: the Process is terminating`)
+	default:
+		v.onShutdown = append(v.onShutdown, fn)
 	}
-	v.onShutdown = append(v.onShutdown, fn)
-	v.lock.Unlock()
+}
+
+func (v *Process) WaitForShutdown() {
+	<-v.done
+}
+
+// ShutdownReason gets the shutdown reason error.
+func ShutdownReason(ctx context.Context) error {
+	v, _ := ctx.Value(ctxShutdownReasonKey).(error)
```

Comment:
This is typical method to set/get context value.
- There is custom private type `ctxKey string`, so no other package can set same key :exclamation: 
- Context value has `any` type, but we know that it is `error`, so private key constant + public getter function is the best approach.